app/ota_client.py

```diff
@@ -896,48 +892,147 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
+            Init the worker pool with shared variables
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        global global_var_dict
+        global_var_dict = gvar_dict
+
+    def _process_regular_files_exit(self, gvar_dict):
+        '''
+            Used by _process_regular_files
+            Update corresponding class attributes
+
+            DO NOT call this method from other functions 
+            except for _process_regular_files
+        '''
+        for k, v in gvar_dict.items():
+            # apply staged data to target attribute
+            vname = k.split("-")
+            if vname[0] == "staging":
+                if vname[1] == "dict":
+                    # TODO: update or replace?
+                    setattr(self, vname[-1], dict(v))
+                elif vname[1] == "list":
+                    setattr(self, vname[-1], list(v))
+
+    def _process_regular_files_log2file_listener(self, flog_path, log_queue, event):
+        '''
+            logger worker for _process_regular_files
+        '''
+        with open(flog_path, "w") as f:
+            while not event.is_set() or not log_queue.empty():
+                    msg = log_queue.get()
+                    f.write(msg+"\n")
+
+    def _process_regular_files(self, rootfs_dir, rfiles_list, target_dir):
+        with Manager() as manager, \
+            tempfile.NamedTemporaryFile(delete=False) as flog:
+            log_file_path, log_q, ev = flog.name, manager.Queue(), manager.Event()
+            # variables passed to child processes
+            # variable naming pattern: <prefix>-<type>-<var_name>
+            # prefix tmp: for temporary use
+            # prefix staging: used to update corresponding class attribute
+            gvar_dict = {
+                "tmp-queue-log_queue": log_q,
+                "tmp-dict-hardlink_reg": manager.dict(), 
+                "staging-dict-_rollback_dict": manager.list()
+            } 
+        
+            # launch log2file logger
+            #   passing file descriptor to sub-process is not possible,
+            #   so we pass the log file path to the logger, and let the logger
+            #   open the log file by itself
+            p_logger = Process(
+                target=self._process_regular_files_log2file_listener, 
+                args=(log_file_path, log_q, ev, )
+                )
+            p_logger.start()
+
+            # default to one worker per CPU core
+            with Pool(
+                initializer=self._process_regular_files_pool_init,
+                initargs=(gvar_dict,)
+            ) as pool:
+                for rfile_inf in rfiles_list:
+                    if int(rfile_inf.links) >=2 and rfile_inf.sha256hash not in gvar_dict["tmp-dict-hardlink_reg"]:
+                        # block the flow until the first copy of hardlinked file is ready
+                        pool.apply(
```

Comment:
(again since outdated)
return value of pool.apply_async should be checked by get?
like:
return value of `pool.apply_async` should be checked by get?
like:
```suggestion
                    results = [pool.apply_async(...) for rfile_inf in rfiles_list]
                    for result in results:
                        result.get()
```